tests/bot/rules/test_links.py

```diff
@@ -0,0 +1,100 @@
+import unittest
+from typing import List, NamedTuple, Tuple
+
+from bot.rules import links
+from tests.helpers import async_test
+
+
+class FakeMessage(NamedTuple):
+    author: str
+    content: str
+
+
+class Case(NamedTuple):
+    recent_messages: List[FakeMessage]
+    relevant_messages: Tuple[FakeMessage]
+    culprit: Tuple[str]
+    total_links: int
+
+
+def msg(author: str, total_links: int) -> FakeMessage:
+    """Makes a message with `total_links` links."""
+    content = " ".join(["https://pydis.com"] * total_links)
+    return FakeMessage(author=author, content=content)
+
+
+class LinksTests(unittest.TestCase):
+    """Tests applying the `links` rule."""
+
+    def setUp(self):
+        self.config = {
+            "max": 2,
+            "interval": 10
+        }
+
+    @async_test
+    async def test_links_within_limit(self):
+        """Messages with an allowed amount of links."""
+        cases = (
+            [msg("bob", 0)],
+            [msg("bob", 2)],
+            [msg("bob", 3)],
+            [msg("bob", 1), msg("bob", 1)],
+            [msg("bob", 3), msg("alice", 3)]
```

Comment:
I don't understand how this case passes the filter if the config has `max` set to 2 rather than 3.

Reply:
Never mind, it's because `messages_with_links > 1`

Reply:
The case can be annotated with an inline comment, do you think that would be helpful?

Reply:
Yes that would actually be nice. There's also a similar situation with line 41 so perhaps move them together and put a comment above.

Reply:
I annotated them separately as I intended the last one to test that the filter will only count links from the latest author. Let me know if you think another way would be better.

Reply:
Then I think it's better to make that last case have only 2 links per message so it has more focus and less points of failure.

Reply:
I agree, this is a good point. Adjusted in latest commit.